src/System.IO.Pipes/tests/NamedPipeTests/NamedPipeTest.Simple.cs

```diff
@@ -540,6 +623,80 @@ public void OperationsOnUnconnectedClient()
+        {
+            using (NamedPipePair pair = CreateNamedPipePair())
+            {
+                NamedPipeClientStream client = pair.clientStream;
+                byte[] buffer = new byte[] { 0, 0, 0, 0 };
+
+                pair.Connect();
+                if (client.CanRead)
+                {
+                    Task clientReadToken = client.ReadAsync(buffer, 0, buffer.Length, CancellationToken.None);
+                    unsafe
+                    {
+                        Assert.True(CancelIoEx(client.SafePipeHandle, null), "Outer cancellation failed");
+                    }
+
+                    await Assert.ThrowsAnyAsync<OperationCanceledException>(() => clientReadToken);
+                    Assert.True(clientReadToken.IsCanceled);
+                }
+                if (client.CanWrite)
+                {
+                    Task clientWriteToken = client.WriteAsync(buffer, 0, buffer.Length, CancellationToken.None);
+                    unsafe
+                    {
+                        Assert.True(CancelIoEx(client.SafePipeHandle, null), "Outer cancellation failed");
+                    }
+
+                    await Assert.ThrowsAnyAsync<OperationCanceledException>(() => clientWriteToken);
+                    Assert.True(clientWriteToken.IsCanceled);
+                }
+            }
+        }
+
+        [Fact]
+        [PlatformSpecific(PlatformID.Windows)]
+        // [ActiveIssue(2640)]
+        public async Task CancelTokenOn_Client_ReadWriteCancelledToken_Throws_OperationCanceledException()
+        {
+            using (NamedPipePair pair = CreateNamedPipePair())
+            {
+                NamedPipeClientStream client = pair.clientStream;
+                byte[] buffer = new byte[] { 0, 0, 0, 0 };
+
+                pair.Connect();
+                if (client.CanRead)
+                {
+                    var cts = new CancellationTokenSource();
+                    Task clientReadToken = client.ReadAsync(buffer, 0, buffer.Length, cts.Token);
+                    unsafe
+                    {
+                        Assert.True(CancelIoEx(client.SafePipeHandle, null), "Outer cancellation failed");
+                    }
+
+                    await Assert.ThrowsAnyAsync<IOException>(() => clientReadToken);
+                    Assert.False(clientReadToken.IsCanceled);
+                }
+                if (client.CanWrite)
+                {
+                    var cts = new CancellationTokenSource();
+                    Task clientWriteToken = client.WriteAsync(buffer, 0, buffer.Length, cts.Token);
+                    unsafe
+                    {
+                        Assert.True(CancelIoEx(client.SafePipeHandle, null), "Outer cancellation failed");
+                    }
+
+                    await Assert.ThrowsAnyAsync<IOException>(() => clientWriteToken);
+                    Assert.False(clientWriteToken.IsCanceled);
```

Comment:
I'm not following.  In what situation could the ```await Assert.ThrowsAnyAsync<IOException>(() => clientWriteToken);`` pass successfully but then the ```Assert.False(clientWriteToken.IsCanceled)``` could fail?  In the code snippet you shared, we'd never get to the Assert.False because the Assert.ThrowsAnyAsync would fail due to the thrown OperationCanceledException not being an IOException.